### DISCUSSION TITLE
Character counter state machine based PX 2013 file parsing

### DISCUSSION THREAD

**Discussion by mikaelhg** [statisticssweden/PxWeb#303] [Q&A]
Hi!

After writing an ANTLR4 [specification](https://github.com/mikaelhg/pxspec/blob/master/px.g4) for PX 2013 to figure out the corner cases of the format, I spent some time thinking about how to balance the need for performance, memory efficiency and maintainability in production parser code,  and came to the conclusion that I could basically skip building a relatively complicated to maintain traditional state machine, and instead make all of the necessary decisions based on character counters which will fit in processor registers if necessary.

I built a Python POC of the concept:   
https://github.com/mikaelhg/pxspec/blob/master/poc_statemachine_2.py#L131

in order to get the rest of the project's nonfunctional requirements from the Go platform, such as easy cross compiling to Linux, Windows and OSX, on AMD64 and ARM. The current Go implementation parses and transforms a 221 MB PX file to CSV in 1.5 seconds, while the Python implementation takes 20 seconds.

At the moment, I didn't yet build in specific `TIMEVAL` or `HIERARCHY` support beyond passing valid header values through the parser.

This kind of an approach will require a second phase in parsing, where the keyword  languages and subkeys are checked against a list of keywords supporting those, and the special header values like integers and timestamps are checked for validity. This is somewhat outside the scope of what I'm trying to accomplish here, but would be easy to implement on the same code bases. 

Any ideas or feedback on what, beyond having to build a more sophisticated way of handling UTF-8 runes if I add support for UTF-8 parsing, could go wrong with such an approach, or any thoughts about corner cases I might have neglected?

**Comment by mikaelhg** [comment]
So far so good. The concept seems to work for this use case. I've ported the parser to Python, Go and Java thus far.

**Comment by likp** [comment]
I am glad to hear that your work proceeds in a promising way.

I am guessing that the io.TextIOWrapper will take care of different encodings since you are reading one character at the time so it will probably work fine with UTF-8 with your approach. But my knowledge in Python is very limited.

What was difficult for us when we wrote the VB.NET parser was to find out the correct encoding of the file since there is no good way to figure it you from the file content itself. We ended up using a C# port of the Mozilla Universal Charset Detector.

**Reply by mikaelhg** [reply]
I'm planning to check every time we hit a `;` character in the header, whether it was a `CODEPAGE`, and if so, switch the character set decoder to one specified in the header. Didn't implement it yet, but I suppose that would be a good next step.

I.e. here: https://github.com/mikaelhg/px2csv-java/blob/40fdfd586f4dee50cab9ed5827608163245acc2d/src/main/java/io/mikael/px2/PxParser.java#L119

**Comment by mikaelhg** [comment]
I came up with a decent solution, and implemented header and dense data processing. 

It takes 1.2 seconds and a few megabytes of memory to convert a 600 MB PX cube into CSV.

I'll check back when I have time to implement sparse data parsing, and convert the standalone utility codebases into separate library and CLI tool repos.
